src/Cache/Context/OgMembershipStateCacheContext.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+namespace Drupal\og\Cache\Context;
+
+use Drupal\Core\Cache\CacheableMetadata;
+use Drupal\Core\Cache\Context\CacheContextInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\og\GroupTypeManager;
+use Drupal\og\MembershipManagerInterface;
+use Drupal\og\OgMembershipInterface;
+
+/**
+ * Defines a cache context service, for "membership state" caching.
+ *
+ * Cache context ID: 'og_membership_state'
+ */
+class OgMembershipStateCacheContext implements CacheContextInterface {
+
+
+  /**
+   * The group type manager service.
+   *
+   * @var \Drupal\og\GroupTypeManager
+   */
+  protected $groupTypeManager;
+
+
+  /**
+   * The membership manager service.
+   *
+   * @var \Drupal\og\MembershipManagerInterface
+   */
+  protected $membershipManager;
+
+  /**
+   * The route match service.
+   *
+   * @var \Drupal\Core\Routing\RouteMatchInterface
+   */
+  protected $routeMatch;
+
+  /**
+   * The current user.
+   *
+   * @var \Drupal\Core\Session\AccountInterface
+   */
+  protected $user;
+
+  /**
+   * Constructs a new UserCacheContextBase class.
+   *
+   * @param \Drupal\Core\Session\AccountInterface $user
+   *   The current user.
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   The current route match service.
+   * @param \Drupal\og\GroupTypeManager $group_type_manager
+   *   The group type manager service.
+   * @param \Drupal\og\MembershipManagerInterface $membership_manager
+   *   The membership manager service.
+   */
+  public function __construct(AccountInterface $user, RouteMatchInterface $route_match, GroupTypeManager $group_type_manager, MembershipManagerInterface $membership_manager) {
+    $this->user = $user;
+    $this->routeMatch = $route_match;
+    $this->groupTypeManager = $group_type_manager;
+    $this->membershipManager = $membership_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getLabel() {
+    return t('OG membership state');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getContext() {
+    if (!$route_contexts = $this->routeMatch->getRouteObject()->getOption('parameters')) {
+      // No "parameters" defined in the route.
+      return 'none';
+    }
```

Comment:
Depending directly on the route parameters severely limits the usefulness of this cache context. It will for example not be possible to use this on group content pages, because they won't match the route of the group to which they belong.

An example: in our project all group content has a header that displays the group to which the content belongs. In this header we have some local actions the user can take (for example join the group if they are not a member yet). So we need to vary by membership state cache context on our group content pages.

I am wondering if we can add the current group to the generic route context.